source/common/router/config_impl.cc

```diff
@@ -1372,23 +1372,55 @@ RouteConstSharedPtr RouteEntryImplBase::pickWeightedCluster(const Http::HeaderMa
     }
   }
 
+  auto runtime_key_prefix_configured =
+      (weighted_clusters_config_->runtime_key_prefix_.length() ? true : false);
+  uint32_t total_cluster_weight = weighted_clusters_config_->total_cluster_weight_;
+  absl::InlinedVector<uint32_t, 4> clusterWeights;
+
+  // if runtime config is used, we need to recompute total_weight
+  if (runtime_key_prefix_configured) {
+    // Temporary storage to hold consistent cluster weights. Since cluster weight
+    // can be changed with runtime keys, we need a way to gather all the weight
+    // and aggregate the total without a change in between.
+    // The InlinedVector will be able to handle at least 4 cluster weights
+    // without allocation. For cases when more clusters are needed, it is
+    // reserved to ensure at most a single allocation.
+    clusterWeights.reserve(weighted_clusters_config_->weighted_clusters_.size());
+
+    total_cluster_weight = 0;
+    for (const WeightedClusterEntrySharedPtr& cluster :
+         weighted_clusters_config_->weighted_clusters_) {
+      auto cluster_weight = cluster->clusterWeight();
+      clusterWeights.push_back(cluster_weight);
+      if (cluster_weight > std::numeric_limits<uint32_t>::max() - total_cluster_weight) {
+        ENVOY_LOG(debug, "Sum of weight cannnot overflow 2^32");
+        return nullptr;
+      }
+      total_cluster_weight += cluster_weight;
+    }
+  }
+
+  if (!total_cluster_weight) {
+    ENVOY_LOG(debug, "Sum of weight cannnot be zero");
+    return nullptr;
+  }
   const uint64_t selected_value =
       (random_value_from_header.has_value() ? random_value_from_header.value() : random_value) %
-      weighted_clusters_config_->total_cluster_weight_;
+      total_cluster_weight;
   uint64_t begin = 0;
   uint64_t end = 0;
+  auto clusterWeight = clusterWeights.begin();
```

Comment:
ditto.` cluster_weight`

Could you please also audit the code change and make sure it is aligned with https://github.com/envoyproxy/envoy/blob/main/STYLE.md, Thanks!

Reply:
fixed